core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -669,16 +668,934 @@ impl CheckedAdd for Fixed64 {
+				// trivial. The latter will not overflow this branch will only happen if the sum of
+				// `u - v - k` part has been negative, hence `u + B - v - k < b`.
+				w.set(j, s as Single);
+			}
+
+			if k.is_zero() {
+				Ok(w)
+			} else {
+				Err(w)
+			}
+		}
+
+		/// Multiplies n-limb number `self` with m-limb number `other`.
+		///
+		/// The resulting number will always have `n + m` limbs.
+		///
+		/// This function does not strip the output and returns the original allocated `n + m`
+		/// limbs. The caller may strip the output if desired.
+		///
+		/// Taken from "The Art of Computer Programming" by D.E. Knuth, vol 2, chapter 4.
+		pub fn mul(self, other: &Self) -> Self {
+			let n = self.len();
+			let m = other.len();
+			let mut w = Self::with_capacity(m + n);
+
+			for j in 0..n {
+				if self.get(j) == 0 {
+					// Note: `with_capacity` allocates with 0. Explicitly set j + m to zero if
+					// otherwise.
+					continue;
+				}
+
+				let mut k = 0;
+				for i in 0..m {
+					// PROOF: (B−1) × (B−1) + (B−1) + (B−1) = B^2 −1 < B^2. addition is safe.
+					let t =
+						mul_single(self.get(j), other.get(i))
+						+ Double::from(w.get(i + j))
+						+ Double::from(k);
+					w.set(i + j, (t % B) as Single);
+					// PROOF: (B^2 - 1) / B < B. conversion is safe.
+					k = (t / B) as Single;
+				}
+				w.set(j + m, k);
+			}
+			w
+		}
+
+		/// Divides `self` by a single limb `other`. This can be used in cases where the original
+		/// division cannot work due to the divisor (`other`) being just one limb.
+		///
+		/// Invariant: `other` cannot be zero.
+		pub fn div_unit(self, mut other: Single) -> Self {
+			other = other.max(1);
+			let n = self.len();
+			let mut out = Self::with_capacity(n);
+			let mut r: Single = 0;
+			// PROOF: (B-1) * B + (B-1) still fits in double
+			let with_r = |x: Double, r: Single| { r as Double * B + x };
+			for d in (0..=n-1).rev() {
+				let (q, rr) = div_single(with_r(self.get(d).into(), r), other) ;
+				out.set(d, q as Single);
+				r = rr;
+			}
+			out
+		}
+
+		/// Divides an `n + m` limb self by a `n` limb `other`. The result is a `m + 1` limb
+		/// quotient and a `n` limb remainder, if enabled by passing `true` in `rem` argument, both
+		/// in the form of an option's `Ok`.
+		///
+		/// - requires `other` to be stripped and have no leading zeros.
+		/// - requires `self` to be stripped and have no leading zeros.
+		/// - requires `other` to have at least two limbs.
+		/// - requires `self` to have a greater or equal length compared to `other`.
+		///
+		/// All arguments are examined without being stripped for the above conditions. If any of
+		/// the above fails, `None` is returned.`
+		///
+		/// Taken from "The Art of Computer Programming" by D.E. Knuth, vol 2, chapter 4.
+		pub fn div(self, other: &Self, rem: bool) -> Option<(Self, Self)> {
+			if other.len() <= 1
+				|| other.msb() == 0
+				|| self.msb() == 0
+				|| self.len() <= other.len()
```

Comment:
is this correct ? 400/300 returns None ?

Reply:
yes -- this algorithm, up to my reading from it only works for when the divisor is strictly larger than dividend Of course, there might be some variations that I have not yet seen or thought of. At least in the book, it is exactly the same. For this, there is a `div_single` defined. One can easily write a wrapper and do something like `if lhs.len() > 1 { rhs.div(lhs) } else { rhs.div_single(lhs) }`. Maybe we can provide this. But I am already doing it manually in `multiply_by_rational`

Reply:
then instead of 
```

		/// - requires `self` to have a greater or equal length compared to `other`.
```

you should write ?
```

		/// - requires `self` to have a greater compared to `other`.
```